### DISCUSSION TITLE
Ignoring an empty line when the parser is able to parse it but I don't want it to.

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-parsing#266] [Q&A] [answered]
This looks very similar to the issue here... https://github.com/pointfreeco/swift-parsing/discussions/260

But my issue is that my parser is parsing an array of digits. So when it sees an empty line at the end of the file it parses it fine and returns an empty array for that line.

I've dealt with this in other files but in this particular file it isn't working because it is able to parse. It's just parsing incorrectly.

The input file is like this...

```
30373
25512
65332
33549
35390

```

(Note the empty last line).

And I want to parse it into a 2D array of Int... `[[Int]]`.

So my parsers looks like this..

```
enum Parsers {
  static let row = Many {
    Digits(1)
  } separator: {
    ""
  }

  static let input = Many {
    row
  } separator: {
    Whitespace(1, .vertical)
  }
}
```

The output I get from this is...
```
[
  [3, 0, 3, 7, 3],
  [2, 5, 5, 1, 2],
  [6, 5, 3, 3, 2],
  [3, 3, 5, 4, 9],
  [3, 5, 3, 9, 0],
  []
]
```

As you can see it is parsing the last empty line as an array of Digits of count 0.

I can't seem to find a way to make this be ignored. The other solutions of using a `terminator` on the Many seems to just throw errors. I guess that's because there is no terminator to deal with. I have parsed all the input. My problem is that I want to not parse the last input line.

Thanks

**Comment by mbrandonw** [comment, marked as answer]
Hi @oliverfoggin, for the row parser I think you want `Many(minimum: 1)`, that way it fails on the last line, leaving it unconsumed. Then you can finish it off with a `terminator` parser that consumes all trailing whitespace. Want to give that a shot and let me know if it works?

**Reply by oliverfoggin** [reply]
Thanks! There doesn't seem to be a `Many(minimum: 1)` syntax. But there is a `range` input. I've updated to this...

```
private enum Parsers {
	static let row = Many(1...) {
		Digits(1)
	} separator: {
		""
	}

	static let input = Many {
		row
	} separator: {
		Whitespace(1, .vertical)
	} terminator: {
		Whitespace(1, .vertical)
	}
}
```

But now it is missing off the penultimate line so I'm missing the line... `[3, 5, 3, 9, 0]`.

**Reply by mbrandonw** [reply]
Are you sure? The test seems to pass for me:

```swift
func testFoo() throws {
  XCTAssertEqual(
    try Parsers.input.parse("""
      30373
      25512
      65332
      33549
      35390

      """),
    [
      [3,0,3,7,3,],
      [2,5,5,1,2,],
      [6,5,3,3,2,],
      [3,3,5,4,9,],
      [3,5,3,9,0,],
    ]
  )
}
```

**Reply by oliverfoggin** [reply]
Sorry, you're right. 🤦🏻

I added a `.dropLast()` after parsing to make the previous version work. And I hadn't removed it yet.

Sorted now though! Thanks 😄
